cpp/src/arrow/compute/kernels/hash_aggregate_numeric.cc

```diff
@@ -162,7 +164,11 @@ struct GroupedReducingAggregator : public GroupedAggregator {
   template <typename T = Type>
   static enable_if_decimal<T, std::shared_ptr<DataType>> GetOutType(
       const std::shared_ptr<DataType>& in_type) {
-    return in_type;
+    if constexpr (PromoteDecimal) {
+      return WidenDecimalToMaxPrecision(in_type).ValueOrDie();
```

Comment:
```suggestion
      return WidenDecimalToMaxPrecision(in_type);
```

Reply:
You might want to keep the error propagation.

Reply:
See below.